waiter/src/waiter/scheduler.clj

```diff
@@ -842,8 +842,54 @@
                                  ([service-id] (retrieve-syncer-state @syncer-state-atom service-id)))}))
 
 ;;
-;; Support for tracking killed instances
+;; Support for tracking killed instance candidates
 ;;
+(defn- cleanup-kill-candidate
+  "Cleans up state for a kill candidate.
+   If candidate was not killed but still failed, we trigger the failed instance callback."
+  [kill-candidates-state instance-id]
+  (let [{:keys [track-failed-instance tracked-states]} (get kill-candidates-state instance-id)]
+    (when (and track-failed-instance
+               (contains? tracked-states :failed)
+               (not (contains? tracked-states :killed)))
+      (log/info "tracking kill candidate as failed instance" {:id instance-id})
```

Comment:
We probably don't want to log inside `swap!` since we could end up calling this multiple times if the swap fails.

From the [docs for `swap!`](https://clojuredocs.org/clojure.core/swap!):

> Note that `f` may be called multiple times, and thus should be free of side effects.  Returns the value that was swapped in.

Can we refactor this to pass in the atom, do the `swap!` inside this function, and then log based on the result?